akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventsBySlicesFirehose.scala

```diff
@@ -0,0 +1,223 @@
+
+    new Firehose(firehoseHub)
+  }
+
+  // FIXME can this be in a ReadJournal
+  def eventsBySlices[Event](
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset): Source[EventEnvelope[Event], NotUsed] = {
+
+    val firehose = getFirehose(entityType, minSlice to maxSlice)
+    val consumerId = UUID.randomUUID().toString
+    val catchupKillSwitch = KillSwitches.shared("catchupKillSwitch")
+    val catchup =
+      eventsBySlices[Event](entityType, minSlice, maxSlice, offset, firehose = false)
+        .map {
+          case env: EventEnvelope[Any] @unchecked => CatchupEventEnvelope(env)
+        }
+        .via(catchupKillSwitch.flow)
+
+    firehose.firehoseHub
+      .map { env =>
+        firehose.updateConsumerTracking(consumerId, timestampOffset(env).timestamp)
+        FirehoseEventEnvelope(env)
+      }
+      .merge(catchup)
+      .via(MergeSelector.flow(consumerId, catchupKillSwitch))
+      .map(_.asInstanceOf[EventEnvelope[Event]])
+  }
+
+  // can be overridden in tests
+  @nowarn("msg=never used") // firehose param used in tests
+  protected def eventsBySlices[Event](
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset,
+      firehose: Boolean): Source[EventEnvelope[Event], NotUsed] = {
+    val queryPluginId = "akka.persistence.r2dbc.query"
+    PersistenceQuery(system)
+      .readJournalFor[EventsBySliceQuery](queryPluginId)
+      .eventsBySlices(entityType, minSlice, maxSlice, offset)
+  }
+
+}
+
+// FIXME maybe write as GraphStage instead
+object MergeSelector {
+  import EventsBySlicesFirehose._
+  private sealed trait Mode
+  private case object CatchUpOnly extends Mode
+  private final case class Both(caughtUpTimestamp: Instant) extends Mode
+  private case object FirehoseOnly extends Mode
+
+  private val log = LoggerFactory.getLogger(classOf[MergeSelector.type])
+
+  def flow(consumerId: String, catchupKillSwitch: KillSwitch): Flow[WrappedEventEnvelope, EventEnvelope[Any], NotUsed] =
+    Flow[WrappedEventEnvelope].statefulMapConcat {
+      var mode: Mode = CatchUpOnly
+      var firehoseOffset: TimestampOffset = TimestampOffset(Instant.EPOCH, Map.empty)
+
+      def updateFirehoseOffset(env: EventEnvelope[Any]): Unit = {
+        val offset = timestampOffset(env)
+        if (offset.timestamp.isAfter(firehoseOffset.timestamp))
+          firehoseOffset = offset // update when newer
+      }
+
+      def isCaughtUp(env: EventEnvelope[Any]): Boolean = {
+        val offset = timestampOffset(env)
+        firehoseOffset.timestamp != Instant.EPOCH && !firehoseOffset.timestamp.isAfter(offset.timestamp)
+      }
+
+      () => {
+        case FirehoseEventEnvelope(env) =>
+          mode match {
+            case CatchUpOnly =>
+              updateFirehoseOffset(env)
+              Nil // skip
+            case Both(_) =>
+              updateFirehoseOffset(env)
+              env :: Nil // emit
```

Comment:
we can add some deduplication, similar to https://github.com/akka/akka-persistence-r2dbc/blob/main/core/src/main/scala/akka/persistence/r2dbc/query/scaladsl/R2dbcReadJournal.scala#L201
but keep in mind the consumer tracking so that firehose events are not filtered out and not reaching the tracking